File: src/Cxx/Widgets/ContourWidget.cxx
#include <vtkCellArray.h>
#include <vtkContourWidget.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkMath.h>
#include <vtkNamedColors.h>
#include <vtkNew.h>
#include <vtkOrientedGlyphContourRepresentation.h>
#include <vtkPoints.h>
#include <vtkPolyData.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>

int main(int /* argc */, char* /* argv */[])
{
  vtkNew<vtkNamedColors> colors;

  // Create the contour widget
  vtkNew<vtkContourWidget> contourWidget;

  // Override the default representation for the contour widget to customize its
  // look
  vtkNew<vtkOrientedGlyphContourRepresentation> contourRepresentation;
  contourRepresentation->GetLinesProperty()->SetColor(
      colors->GetColor3d("Red").GetData());
  contourWidget->SetRepresentation(contourRepresentation);

  // Generate a set of points arranged in a circle
  int numPts = 10;
  vtkNew<vtkPoints> points;
  for (int i = 0; i < numPts; i++)
  {
    // Create numPts points evenly spread around a circumference of radius 0.1
    const double angle = 2.0 * vtkMath::Pi() * i / numPts;
    points->InsertPoint(static_cast<vtkIdType>(i), 0.1 * cos(angle),
                        0.1 * sin(angle), 0.0);
  }

  // Create a cell array to connect the points into meaningful geometry
  vtkIdType* vertexIndices = new vtkIdType[numPts + 1];
  for (int i = 0; i < numPts; i++)
  {
    vertexIndices[i] = static_cast<vtkIdType>(i);
  }
  // Set the last vertex to 0; this means the last line segment will join the
  // 19th point (vertices[19]) with the first one (vertices[0]), thus closing
  // the circle.
  vertexIndices[numPts] = 0;
  vtkNew<vtkCellArray> lines;
  lines->InsertNextCell(numPts + 1, vertexIndices);

  // Create polydata to hold the geometry just created, and populate it
  vtkNew<vtkPolyData> polydata;
  polydata->SetPoints(points);
  polydata->SetLines(lines);

  // Create the renderer to visualize the scene
  vtkNew<vtkRenderer> renderer;
  renderer->SetBackground(colors->GetColor3d("MidnightBlue").GetData());

  // Create the GUI window to hold the rendered scene
  vtkNew<vtkRenderWindow> renderWindow;
  renderWindow->AddRenderer(renderer);
  renderWindow->SetWindowName("ContourWidget");
  renderWindow->SetSize(600, 600);

  // Create the events manager for the renderer window
  vtkNew<vtkRenderWindowInteractor> interactor;
  interactor->SetRenderWindow(renderWindow);

  // Use the "trackball camera" interactor style, rather than the default
  // "joystick camera"
  vtkNew<vtkInteractorStyleTrackballCamera> style;
  interactor->SetInteractorStyle(style);

  // Set up the contour widget within the visualization pipeline just assembled
  contourWidget->SetInteractor(interactor);
  contourWidget->On(); // Turn on the interactor observer
  contourWidget->Initialize(polydata);
  renderer->ResetCamera(); // Reposition camera to fit the scene elements

  // Start the interaction
  renderWindow->Render();
  interactor->Start();

  return EXIT_SUCCESS;
}

File: src/Python/Widgets/ContourWidget.py
#!/usr/bin/python

import math
import sys

# noinspection PyUnresolvedReferences
import vtkmodules.vtkInteractionStyle
# noinspection PyUnresolvedReferences
import vtkmodules.vtkRenderingOpenGL2
from vtkmodules.vtkCommonColor import vtkNamedColors
from vtkmodules.vtkCommonCore import (
    vtkCommand,
    vtkPoints
)
from vtkmodules.vtkCommonDataModel import (
    vtkCellArray,
    vtkPolyData
)
from vtkmodules.vtkInteractionWidgets import (
    vtkContourWidget,
    vtkOrientedGlyphContourRepresentation,
    vtkWidgetEvent
)
from vtkmodules.vtkRenderingCore import (
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkRenderer
)


def main():
    colors = vtkNamedColors()

    # colors.SetColor('bkg', [0.1, 0.2, 0.4, 1.0])

    # Create the RenderWindow, Renderer and both Actors

    renderer = vtkRenderer()
    renderer.SetBackground(colors.GetColor3d('MidnightBlue'))

    renderWindow = vtkRenderWindow()
    renderWindow.AddRenderer(renderer)
    renderWindow.SetWindowName('ContourWidget')
    renderWindow.SetSize(600, 600)

    interactor = vtkRenderWindowInteractor()
    interactor.SetRenderWindow(renderWindow)

    contourRep = vtkOrientedGlyphContourRepresentation()
    contourRep.GetLinesProperty().SetColor(colors.GetColor3d('Red'))

    contourWidget = vtkContourWidget()
    contourWidget.SetInteractor(interactor)
    contourWidget.SetRepresentation(contourRep)
    contourWidget.On()

    for arg in sys.argv:
        if '-Shift' == arg:
            contourWidget.GetEventTranslator().RemoveTranslation(
                vtkCommand.LeftButtonPressEvent)
            contourWidget.GetEventTranslator().SetTranslation(
                vtkCommand.LeftButtonPressEvent,
                vtkWidgetEvent.Translate)
        elif '-Scale' == arg:
            contourWidget.GetEventTranslator().RemoveTranslation(
                vtkCommand.LeftButtonPressEvent)
            contourWidget.GetEventTranslator().SetTranslation(
                vtkCommand.LeftButtonPressEvent,
                vtkWidgetEvent.Scale)

    pd = vtkPolyData()

    points = vtkPoints()

    num_pts = 21
    for i in range(0, num_pts):
        angle = 2.0 * math.pi * i / 20.0
        points.InsertPoint(i, 0.1 * math.cos(angle),
                           0.1 * math.sin(angle), 0.0)
        # lines.InsertNextCell(i)
    vertex_indices = list(range(0, num_pts))
    vertex_indices.append(0)
    lines = vtkCellArray()
    lines.InsertNextCell(num_pts + 1, vertex_indices)

    pd.SetPoints(points)
    pd.SetLines(lines)

    # contourWidget.Initialize(pd, 1)
    contourWidget.Initialize(pd, 1)
    contourWidget.Render()
    renderer.ResetCamera()
    renderWindow.Render()

    interactor.Initialize()
    interactor.Start()


if __name__ == '__main__':
    main()
